test/diff/helper.go

```diff
@@ -0,0 +1,44 @@
+/*
+Copyright 2020 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package diff
+
+import (
+	"github.com/google/go-cmp/cmp"
+	"testing"
+)
+
+// ErrorIfWantNotGot Wrap cmp.Diff with a t.Helper(). It takes the testing.T
+// want,got, format message and cmp.option as an argument. Therefore, if diff was found
+// it will call t.Error of format and diff that wrapped by PrintWantGot func
+func ErrorIfWantNotGot(t *testing.T, want interface{}, got interface{}, format string, opts ...cmp.Option) {
+	// t.Helper() used to marks the calling function as a test helper function
+	t.Helper()
+	if d := cmp.Diff(want, got, opts...); d != "" {
+		t.Errorf(format, PrintWantGot(d))
```

Comment:
Here's a test that prints a diff currently in our codebase:

```go
for _, tt := range tests {
        t.Run(tt.name, func(t *testing.T) {
                got := resources.ApplyParameters(tt.args.ts, tt.args.tr, tt.args.dp...)
                if d := cmp.Diff(tt.want, got); d != "" {
                        t.Errorf("ApplyParameters() got diff %s", diff.PrintWantGot(d))
                }
        })
}
```

And here's what it would look like with this change:

```go
for _, tt := range tests {
        t.Run(tt.name, func(t *testing.T) {
                 got := resources.ApplyParameters(tt.args.ts, tt.args.tr, tt.args.dp...)
                 diff.ErrorIfWantNotGot(t, tt.want, got, "ApplyParameters() got diff %s")
        })
}
```

Notice that I must still include `"%s"`, otherwise the diff won't be printed.  I think it might be less surprising if the string didn't have to include `%s`?  So I could write:

```go
diff.ErrorIfWantNotGot(t, tt.want, got, "ApplyParameters() diff")
```

My suggestion is to accept a string message but don't use it for formatting.  Instead update this line to:

```go
t.Errorf(fmt.Sprintf("%s %s", message, PrintWantGot(d))
```

Same for `FatalIfWantNotGot`.  What do you think of this idea?

Reply:
What about:
```
func ErrorOnDiff(t *testing.T, name string, want interface{}, got interface{}, opts ...cmp.Option)
```

and 

```
diff.ErrorOnDiff(t, "ApplyParameters()", tt.want, got)
```

Reply:
I think it would be nice if the name of the func made explicit the order of the params.  Want/Got, just like we have diff.PrintWantGot.  I'm not too fussed about what the overall names are but I do think it's worth preserving that pattern.

Reply:
@sbwsg thanks for correcting me 
the explanation is very helpful
I've already updated the code